File: regimen_manager.py

import json
import os
from typing import Dict, List, Optional, Callable

class RegimenManager:
    def __init__(self, timer_manager, output_callback: Optional[Callable[[str], None]] = None):
        self.timer_manager = timer_manager
        self.output_callback = output_callback
        self.regimens_file = "regimens.json"
        self.current_regimen = None
        self.current_regimen_queue = []
        self.current_regimen_name = ""
        
    def _print(self, message: str):
        """Print message using callback if available"""
        if self.output_callback:
            self.output_callback(message)
        else:
            print(message)
    
    def load_regimens(self) -> Dict:
        """Load regimens from file"""
        if not os.path.exists(self.regimens_file):
            return {}
        try:
            with open(self.regimens_file, 'r') as f:
                return json.load(f)
        except Exception as e:
            self._print(f"Error loading regimens: {str(e)}")
            return {}
    
    def save_regimens(self, regimens: Dict):
        """Save regimens to file"""
        try:
            with open(self.regimens_file, 'w') as f:
                json.dump(regimens, f, indent=2)
        except Exception as e:
            self._print(f"Error saving regimens: {str(e)}")
    
    def create_regimen(self, name: str, timers: List[Dict]):
        """Create a new regimen"""
        regimens = self.load_regimens()
        regimens[name] = timers
        self.save_regimens(regimens)
        self._print(f"Created regimen '{name}' with {len(timers)} timer(s)")
    
    def list_regimens(self):
        """List all available regimens"""
        regimens = self.load_regimens()
        if not regimens:
            self._print("No regimens available")
            return
        
        self._print("Available regimens:")
        for name, timers in regimens.items():
            timer_list = ", ".join([f"{t['name']} ({t['duration']}s)" for t in timers])
            self._print(f"- {name}: {timer_list}")
    
    def run_regimen(self, name: str):
        """Start a regimen - creates first timer in sequence"""
        regimens = self.load_regimens()
        if name not in regimens:
            self._print(f"Regimen '{name}' not found")
            return
        
        self.current_regimen_name = name
        self.current_regimen_queue = regimens[name].copy()
        
        if not self.current_regimen_queue:
            self._print(f"Regimen '{name}' is empty")
            return
        
        # Start the first timer in the regimen
        self._start_next_timer()
    
    def _start_next_timer(self):
        """Start the next timer in the current regimen queue"""
        if not self.current_regimen_queue:
            self._print(f"Regimen '{self.current_regimen_name}' completed!")
            self.current_regimen_name = ""
            return
        
        next_timer = self.current_regimen_queue.pop(0)
        timer_name = f"{self.current_regimen_name}_{next_timer['name']}"
        
        self._print(f"Starting regimen timer: {timer_name} ({next_timer['duration']}s)")
        
        # Create and start the timer
        self.timer_manager.create_timer(timer_name, next_timer['duration'])
    
    def on_timer_complete(self, timer_name: str):
        """Called when a timer completes - check if it's part of a regimen"""
        if not self.current_regimen_name:
            return
        
        # Check if this timer is part of the current regimen
        if timer_name.startswith(f"{self.current_regimen_name}_"):
            # Start the next timer in the regimen after a short delay
            import threading
            import time
            
            def delayed_start():
                time.sleep(1)  # Brief pause between timers
                self._start_next_timer()
            
            thread = threading.Thread(target=delayed_start)
            thread.daemon = True
            thread.start()

File: src/js/regimen-manager.js
class RegimenManager {
    constructor(timerManager, outputCallback = null) {
        this.timerManager = timerManager;
        this.outputCallback = outputCallback;
        this.regimensFile = "regimens.json";
        this.currentRegimen = null;
        this.currentRegimenQueue = [];
        this.currentRegimenName = "";
    }

    _print(message) {
        if (this.outputCallback) {
            this.outputCallback(message);
        } else {
            console.log(message);
        }
    }

    async loadRegimens() {
        try {
            // Use file-based persistence through Electron IPC
            if (window.electronAPI) {
                const result = await window.electronAPI.loadFile(this.regimensFile);
                if (result.success && result.data) {
                    return result.data;
                }
            }
            return {};
        } catch (error) {
            this._print(`Error loading regimens: ${error.message}`);
            return {};
        }
    }

    async saveRegimens(regimens) {
        try {
            // Use file-based persistence through Electron IPC
            if (window.electronAPI) {
                await window.electronAPI.saveFile(this.regimensFile, regimens);
            }
        } catch (error) {
            this._print(`Error saving regimens: ${error.message}`);
        }
    }

    async createRegimen(name, timers) {
        const regimens = await this.loadRegimens();
        regimens[name] = timers;
        await this.saveRegimens(regimens);
        this._print(`Created regimen '${name}' with ${timers.length} timer(s)`);
    }

    async listRegimens() {
        const regimens = await this.loadRegimens();
        if (Object.keys(regimens).length === 0) {
            this._print("No regimens available");
            return;
        }

        this._print("Available regimens:");
        for (const [name, timers] of Object.entries(regimens)) {
            const timerList = timers.map(t => `${t.name} (${t.duration}s)`).join(", ");
            this._print(`- ${name}: ${timerList}`);
        }
    }

    async runRegimen(name) {
        const regimens = await this.loadRegimens();
        if (!(name in regimens)) {
            this._print(`Regimen '${name}' not found`);
            return;
        }

        this.currentRegimenName = name;
        this.currentRegimenQueue = [...regimens[name]]; // Create a copy

        if (this.currentRegimenQueue.length === 0) {
            this._print(`Regimen '${name}' is empty`);
            return;
        }

        // Start the first timer in the regimen
        this._startNextTimer();
    }

    _startNextTimer() {
        if (this.currentRegimenQueue.length === 0) {
            this._print(`Regimen '${this.currentRegimenName}' completed!`);
            this.currentRegimenName = "";
            return;
        }

        const nextTimer = this.currentRegimenQueue.shift();
        const timerName = `${this.currentRegimenName}_${nextTimer.name}`;

        this._print(`Starting regimen timer: ${timerName} (${nextTimer.duration}s)`);

        // Create and start the timer
        this.timerManager.createTimer(timerName, nextTimer.duration);
    }

    onTimerComplete(timerName) {
        if (!this.currentRegimenName) {
            return;
        }

        // Check if this timer is part of the current regimen
        if (timerName.startsWith(`${this.currentRegimenName}_`)) {
            // Start the next timer in the regimen after a short delay
            setTimeout(() => {
                this._startNextTimer();
            }, 1000); // Brief pause between timers
        }
    }

    // Get regimens for UI display
    async getRegimens() {
        return await this.loadRegimens();
    }

    // Delete a regimen
    async deleteRegimen(name) {
        const regimens = await this.loadRegimens();
        if (name in regimens) {
            delete regimens[name];
            await this.saveRegimens(regimens);
            this._print(`Deleted regimen '${name}'`);
            return true;
        } else {
            this._print(`Regimen '${name}' not found`);
            return false;
        }
    }

    // Check if a regimen exists
    async regimenExists(name) {
        const regimens = await this.loadRegimens();
        return name in regimens;
    }

    // Get regimen details
    async getRegimen(name) {
        const regimens = await this.loadRegimens();
        return regimens[name] || null;
    }

    // Stop current regimen
    stopCurrentRegimen() {
        if (this.currentRegimenName) {
            this._print(`Stopping regimen '${this.currentRegimenName}'`);
            this.currentRegimenName = "";
            this.currentRegimenQueue = [];
        }
    }

    // Get current regimen status
    getCurrentRegimenStatus() {
        if (!this.currentRegimenName) {
            return null;
        }

        return {
            name: this.currentRegimenName,
            remaining: this.currentRegimenQueue.length,
            total: this.getRegimen(this.currentRegimenName)?.length || 0
        };
    }
}

// Make RegimenManager available globally
window.RegimenManager = RegimenManager;